src/datumaro/experimental/schema.py

```diff
@@ -188,5 +280,96 @@ def with_categories(self, categories: Dict[str, "Categories"]) -> "Schema":
                 new_schema.attributes[attr_name].categories = category
             else:
                 raise ValueError(f"Attribute '{attr_name}' not found in schema")
-
         return new_schema
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this Schema to a JSON-compatible dictionary.
+
+        Returns:
+            Dictionary representation of this Schema instance
+        """
+
+        attributes = {}
+        categories = {}
+
+        for name, attr_info in self.attributes.items():
+            # Store both module and name for better type reconstruction
+            type_info = (
+                attr_info.type.__name__
+                if hasattr(attr_info.type, "__name__")
+                else str(attr_info.type)
+            )
+            type_module = (
+                attr_info.type.__module__ if hasattr(attr_info.type, "__module__") else None
+            )
+
+            attributes[name] = {
+                "type": type_info,
+                "type_module": type_module,
+                "field": attr_info.annotation.to_dict(),
+            }
+            if attr_info.categories is not None:
+                categories[name] = attr_info.categories.to_dict()
+
+        return {
+            "attributes": attributes,
+            "categories": categories,
+        }
+
+    @classmethod
+    def from_dict(cls, schema_dict: Dict[str, Any]) -> "Schema":
+        """
+        Deserialize a Schema from a JSON dictionary.
+
+        Args:
+            schema_dict: Dictionary containing schema data
+
+        Returns:
+            Reconstructed Schema instance
+        """
+
+        attributes = {}
+
+        for name, attr_dict in schema_dict["attributes"].items():
+            field = Field.from_dict(attr_dict["field"])
+
+            # Get categories if present
+            categories = None
+            if name in schema_dict.get("categories", {}):
+                categories = Categories.from_dict(schema_dict["categories"][name])
+
+            # Try to reconstruct the type
+            type_name = attr_dict.get("type", "object")
+            type_module = attr_dict.get("type_module")
+
+            # Attempt to reconstruct the actual type
+            if type_module and type_module != "builtins":
+                try:
+                    module = importlib.import_module(type_module)
+                    attr_type = getattr(module, type_name, object)
+                except (ImportError, AttributeError):
+                    attr_type = object
+            elif type_name in ["int", "float", "str", "bool", "list", "dict", "tuple"]:
```

Comment:
You can use `type_name in BUILTIN_TYPES` if you put the constant declaration out of the loop